File: backend-rust/backend-core/src/software_scanner.rs
use serde::{Deserialize, Serialize};
use std::collections::HashMap;
use std::path::Path;
use std::fs;
use chrono::Utc;
use winreg::enums::*;
use winreg::RegKey;

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SoftwareInfo {
    pub name: String,
    pub exec_path: String,
    pub icon_path: String,
    pub category: String,
    pub description: String,
    pub source: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SoftwareCache {
    pub scanned_at: String,
    pub machine_id: String,
    pub count: usize,
    pub software: Vec<SoftwareInfo>,
}

fn get_machine_id() -> String {
    std::env::var("COMPUTERNAME").unwrap_or_else(|_| {
        std::env::var("HOSTNAME").unwrap_or_else(|_| "unknown".to_string())
    })
}

const CATEGORY_KEYWORDS: &[(&str, &[&str])] = &[
    ("browser", &["browser", "chrome", "edge", "firefox", "safari", "opera", "chromium", "brave", "iexplore"]),
    ("editor", &["code", "sublime", "vim", "neovim", "notepad", "emacs", "atom", "brackets", "text", "ide"]),
    ("terminal", &["terminal", "cmd", "powershell", "windowsterminal", "conemu", "cmder", "hyper", "alacritty", "putty", "ssh", "warp"]),
    ("office", &["word", "excel", "powerpoint", "outlook", "onenote", "access", "libreoffice", "wps", "office"]),
    ("image", &["photoshop", "gimp", "krita", "paint", "illustrator", "figma", "sketch", "lightroom", "canvas", "photo", "draw", "inkscape", "corel"]),
    ("video", &["premiere", "after effects", "davinci", "final cut", "vegas", "kdenlive", "shotcut", "obs", "handbrake", "vlc", "mpv", "media player", "potplayer"]),
    ("audio", &["audacity", "ableton", "fl studio", "cubase", "logic", "reason", "studio one", "reaper"]),
    ("development", &["git", "python", "node", "npm", "docker", "kubernetes", "vscode", "visual studio", "intellij", "pycharm", "webstorm", "goland", "clion", "eclipse", "android studio", "xcode", "jupyter", "anaconda", "cmake", "mingw", "gradle", "maven"]),
    ("database", &["mysql", "postgresql", "mongodb", "sqlite", "redis", "oracle", "sql server", "dbeaver", "sequel pro", "heidisql", "navicat", "pgadmin"]),
    ("design", &["blender", "maya", "3ds max", "cinema 4d", "unity", "unreal", "godot", "fusion 360", "autocad", "solidworks", "sketchup"]),
    ("communication", &["discord", "slack", "teams", "zoom", "skype", "telegram", "whatsapp", "wechat", "qq", "dingtalk", "feishu"]),
    ("utility", &["calculator", "calendar", "clock", "notes", "evernote", "notion", "obsidian", "1password", "lastpass", "dropbox", "onedrive"]),
    ("game", &["steam", "epic", "battle.net", "origin", "ubisoft", "gog", "minecraft", "league", "dota", "counter-strike"]),
    ("music", &["spotify", "netease", "qq music", "apple music", "foobar", "winamp", "aimp", "musicbee"]),
    ("pdf", &["acrobat", "reader", "foxit", "sumatra", "pdf reader", "pdf viewer"]),
    ("compression", &["winrar", "7-zip", "winzip", "peazip", "bandizip", "tar", "gzip"]),
    ("download", &["aria2", "motrix", "xdown", "idm", "internet download manager", "transmission", "qbittorrent", "utorrent", "thunder"]),
];

const EXCLUDED_NAMES: &[&str] = &[
    "uninstall", "update", "setup", "install", "readme", "help",
    "documentation", "support", "feedback", "report", "get started",
    "what's new", "about", "license", "release notes",
];

const EXCLUDED_DIRS: &[&str] = &[
    "windows", "windows.old", "system32", "syswow64", "msocache",
    "$recycle.bin", "system volume information", "perflogs",
];

const COMMON_EXES: &[&str] = &[
    "chrome.exe", "firefox.exe", "msedge.exe", "code.exe",
    "notepad++.exe", "sublime_text.exe", "terminal.exe",
    "wt.exe", "powershell.exe", "cmd.exe",
    "spotify.exe", "discord.exe", "slack.exe",
    "obs64.exe", "obs32.exe", "vlc.exe",
    "python.exe", "pythonw.exe",
    "winrar.exe", "7zFM.exe", "sumatrapdf.exe",
];

fn classify_software(name: &str, exec_path: &str) -> String {
    let combined = format!("{} {}", name.to_lowercase(), exec_path.to_lowercase());

    for &(cat, keywords) in CATEGORY_KEYWORDS {
        for kw in keywords {
            if combined.contains(kw) {
                return cat.to_string();
            }
        }
    }
    "other".to_string()
}

fn is_excluded(name: &str) -> bool {
    let lower = name.to_lowercase().trim().to_string();
    EXCLUDED_NAMES.iter().any(|exc| lower.contains(exc))
}

fn scan_start_menu() -> Vec<SoftwareInfo> {
    let mut found = Vec::new();
    let mut seen = std::collections::HashSet::new();

    let start_menu_dirs = vec![
        format!("{}\\Microsoft\\Windows\\Start Menu\\Programs",
            std::env::var("ProgramData").unwrap_or_default()),
        format!("{}\\Microsoft\\Windows\\Start Menu\\Programs",
            std::env::var("APPDATA").unwrap_or_default()),
    ];

    for base_dir in &start_menu_dirs {
        let dir = Path::new(base_dir);
        if !dir.is_dir() {
            continue;
        }
        for entry in walkdir::WalkDir::new(dir)
            .into_iter()
            .filter_map(|e| e.ok())
        {
            let path = entry.path();
            let path_lower = path.display().to_string().to_lowercase();
            if EXCLUDED_DIRS.iter().any(|exc| path_lower.contains(exc)) {
                continue;
            }
            if !path.extension().map(|e| e == "lnk").unwrap_or(false) {
                continue;
            }
            let name = path.file_stem()
                .and_then(|n| n.to_str())
                .unwrap_or("");
            if name.is_empty() || is_excluded(name) {
                continue;
            }
            let lower = name.to_lowercase();
            if seen.contains(&lower) {
                continue;
            }
            seen.insert(lower);

            let category = classify_software(name, &path.display().to_string());
            found.push(SoftwareInfo {
                name: name.to_string(),
                exec_path: path.display().to_string(),
                icon_path: path.display().to_string(),
                category,
                description: format!("从开始菜单发现: {}", name),
                source: "start_menu".to_string(),
            });
        }
    }
    found
}

fn scan_registry() -> Vec<SoftwareInfo> {
    let mut found = Vec::new();
    let mut seen_names = std::collections::HashSet::new();

    let registry_paths = [
        (HKEY_LOCAL_MACHINE, r"SOFTWARE\Microsoft\Windows\CurrentVersion\Uninstall"),
        (HKEY_LOCAL_MACHINE, r"SOFTWARE\WOW6432Node\Microsoft\Windows\CurrentVersion\Uninstall"),
        (HKEY_CURRENT_USER, r"SOFTWARE\Microsoft\Windows\CurrentVersion\Uninstall"),
    ];

    for &(hkey, key_path) in &registry_paths {
        let key = match RegKey::predef(hkey).open_subkey(key_path) {
            Ok(k) => k,
            Err(_) => continue,
        };

        for name_result in key.enum_keys().filter_map(|r| r.ok()) {
            let subkey = match key.open_subkey(&name_result) {
                Ok(k) => k,
                Err(_) => continue,
            };

            let display_name: Option<String> = subkey.get_value("DisplayName").ok();
            let display_name = match display_name {
                Some(ref n) if !n.is_empty() && !is_excluded(n) => n.clone(),
                _ => continue,
            };

            let lower = display_name.to_lowercase();
            if seen_names.contains(&lower) {
                continue;
            }
            seen_names.insert(lower);

            let exec_path: String = subkey.get_value("DisplayIcon").ok()
                .or_else(|| subkey.get_value("InstallLocation").ok())
                .unwrap_or_default();

            let publisher: String = subkey.get_value("Publisher").ok()
                .unwrap_or_default();

            let category = classify_software(&display_name, &exec_path);
            found.push(SoftwareInfo {
                name: display_name.clone(),
                exec_path: exec_path.clone(),
                icon_path: exec_path,
                category,
                description: format!("已安装: {}{}",
                    display_name,
                    if publisher.is_empty() { String::new() } else { format!(" (by {})", publisher) }),
                source: "registry".to_string(),
            });
        }
    }
    found
}

fn scan_common_paths() -> Vec<SoftwareInfo> {
    let mut found = Vec::new();
    let mut seen_names = std::collections::HashSet::new();

    let common_dirs = vec![
        std::env::var("ProgramFiles").unwrap_or_else(|_| r"C:\Program Files".to_string()),
        std::env::var("ProgramFiles(x86)").unwrap_or_else(|_| r"C:\Program Files (x86)".to_string()),
        format!("{}\\Programs", std::env::var("LOCALAPPDATA").unwrap_or_default()),
    ];

    for base_dir in &common_dirs {
        let dir = Path::new(base_dir);
        if !dir.is_dir() {
            continue;
        }

        for entry in walkdir::WalkDir::new(dir)
            .max_depth(4)
            .into_iter()
            .filter_map(|e| e.ok())
        {
            let path = entry.path();
            let path_lower = path.display().to_string().to_lowercase();
            if EXCLUDED_DIRS.iter().any(|exc| path_lower.contains(exc)) {
                continue;
            }
            let fname = path.file_name()
                .and_then(|n| n.to_str())
                .unwrap_or("")
                .to_lowercase();

            if !COMMON_EXES.contains(&fname.as_str()) {
                continue;
            }

            let name = Path::new(&fname)
                .file_stem()
                .and_then(|n| n.to_str())
                .unwrap_or(&fname);

            if is_excluded(name) {
                continue;
            }
            let lower = name.to_lowercase();
            if seen_names.contains(&lower) {
                continue;
            }
            seen_names.insert(lower);

            let category = classify_software(name, &path.display().to_string());
            found.push(SoftwareInfo {
                name: name.to_string(),
                exec_path: path.display().to_string(),
                icon_path: path.display().to_string(),
                category,
                description: format!("在安装目录发现: {}", name),
                source: "common_paths".to_string(),
            });
        }
    }
    found
}

fn merge_software_lists(lists: Vec<Vec<SoftwareInfo>>) -> Vec<SoftwareInfo> {
    let mut seen: HashMap<String, SoftwareInfo> = HashMap::new();

    for list in lists {
        for sw in list {
            let key = sw.name.to_lowercase();
            if let Some(existing) = seen.get(&key) {
                if existing.exec_path.is_empty() && !sw.exec_path.is_empty() {
                    seen.insert(key, sw);
                }
            } else {
                seen.insert(key, sw);
            }
        }
    }

    let mut result: Vec<SoftwareInfo> = seen.into_values().collect();
    result.sort_by(|a, b| a.category.cmp(&b.category).then(a.name.cmp(&b.name)));
    result
}

pub fn scan_all_software() -> Vec<SoftwareInfo> {
    log::info!("开始扫描电脑上的软件...");

    let start_menu = scan_start_menu();
    let registry = scan_registry();
    let common = scan_common_paths();

    let merged = merge_software_lists(vec![start_menu, registry, common]);

    log::info!("扫描完成! 共发现 {} 个软件", merged.len());
    merged
}

pub fn save_software_cache(memory_path: &Path, software_list: &[SoftwareInfo]) {
    fs::create_dir_all(memory_path).ok();
    let cache = SoftwareCache {
        scanned_at: Utc::now().to_rfc3339(),
        machine_id: get_machine_id(),
        count: software_list.len(),
        software: software_list.to_vec(),
    };

    let cache_path = memory_path.join("software_cache.json");
    if let Ok(content) = serde_json::to_string_pretty(&cache) {
        fs::write(&cache_path, content).ok();
    }

    let flag_path = memory_path.join(".software_scanned");
    fs::write(&flag_path, Utc::now().to_rfc3339()).ok();

    log::info!("已缓存 {} 个软件到 {:?}", software_list.len(), cache_path);
}

pub fn load_software_cache(memory_path: &Path) -> Vec<SoftwareInfo> {
    let cache_path = memory_path.join("software_cache.json");
    if !cache_path.exists() {
        return Vec::new();
    }
    match fs::read_to_string(&cache_path) {
        Ok(content) => {
            match serde_json::from_str::<SoftwareCache>(&content) {
                Ok(cache) => {
                    log::info!("从缓存加载了 {} 个软件", cache.software.len());
                    cache.software
                }
                Err(e) => {
                    log::warn!("缓存加载失败: {}", e);
                    Vec::new()
                }
            }
        }
        Err(_) => Vec::new(),
    }
}

pub fn is_software_scanned(memory_path: &Path) -> bool {
    memory_path.join(".software_scanned").exists()
}

pub fn search_software(query: &str, software_list: &[SoftwareInfo], top_k: usize) -> Vec<SoftwareInfo> {
    let query_lower = query.to_lowercase();
    let mut scored: Vec<(i32, &SoftwareInfo)> = Vec::new();

    for sw in software_list {
        let mut score = 0;
        if query_lower.len() >= 2 {
            if sw.name.to_lowercase().contains(&query_lower) {
                score += 10;
            }
            if sw.category.contains(&query_lower) {
                score += 5;
            }
            if sw.description.to_lowercase().contains(&query_lower) {
                score += 3;
            }
            if sw.exec_path.to_lowercase().contains(&query_lower) {
                score += 2;
            }
            for keyword in query_lower.split_whitespace() {
                if sw.name.to_lowercase().contains(keyword) {
                    score += 2;
                }
            }
        }
        if score > 0 {
            scored.push((score, sw));
        }
    }

    scored.sort_by(|a, b| b.0.cmp(&a.0));
    scored.into_iter().take(top_k).map(|(_, sw)| sw.clone()).collect()
}

pub fn get_all_categories(software_list: &[SoftwareInfo]) -> HashMap<String, Vec<String>> {
    let mut cats: HashMap<String, Vec<String>> = HashMap::new();
    for sw in software_list {
        cats.entry(sw.category.clone())
            .or_default()
            .push(sw.name.clone());
    }
    cats
}

pub fn get_software_by_category<'a>(software_list: &'a [SoftwareInfo], category: &str) -> Vec<&'a SoftwareInfo> {
    software_list.iter().filter(|sw| sw.category == category).collect()
}

pub fn scan_and_cache(memory_path: &Path) -> Vec<SoftwareInfo> {
    if is_software_scanned(memory_path) {
        // 加载缓存并检查设备是否变更
        let cached = load_software_cache(memory_path);
        let current_machine = get_machine_id();

        if !cached.is_empty() {
            // 读取缓存中的 machine_id
            let cache_path = memory_path.join("software_cache.json");
            let cached_machine = fs::read_to_string(&cache_path)
                .ok()
                .and_then(|s| serde_json::from_str::<SoftwareCache>(&s).ok())
                .map(|c| c.machine_id)
                .unwrap_or_default();

            if cached_machine == current_machine {
                // 同一设备：快速增量扫描（更新增删）
                log::info!("同一设备 ({}), 执行快速增量扫描...", current_machine);
                let updated = quick_rescan(memory_path, &cached);
                if updated.len() != cached.len() {
                    log::info!("增量扫描完成: {} -> {} 个软件", cached.len(), updated.len());
                } else {
                    log::info!("增量扫描完成: 无变化 ({} 个软件)", updated.len());
                }
                return updated;
            } else {
                // 新设备：重置缓存，全量扫描
                log::info!("检测到新设备 ({} -> {}), 重新全量扫描...",
                    if cached_machine.is_empty() { "无" } else { &cached_machine },
                    current_machine);
                clear_software_cache(memory_path);
            }
        }
    }

    let software_list = scan_all_software();
    save_software_cache(memory_path, &software_list);
    software_list
}

fn clear_software_cache(memory_path: &Path) {
    let cache_path = memory_path.join("software_cache.json");
    let flag_path = memory_path.join(".software_scanned");
    let _ = fs::remove_file(&cache_path);
    let _ = fs::remove_file(&flag_path);
    log::info!("已清除软件缓存");
}

fn quick_rescan(memory_path: &Path, cached: &[SoftwareInfo]) -> Vec<SoftwareInfo> {
    // 快速扫描：开始菜单 + 注册表（跳过深度目录遍历）
    let start_menu = scan_start_menu();
    let registry = scan_registry();
    let quick_new = merge_software_lists(vec![start_menu, registry]);

    let quick_map: HashMap<String, &SoftwareInfo> = quick_new.iter()
        .map(|sw| (sw.name.to_lowercase(), sw))
        .collect();

    let mut result: Vec<SoftwareInfo> = Vec::new();
    let mut removed_count: usize = 0;

    // 保留仍然存在且路径有效的软件
    for sw in cached {
        let key = sw.name.to_lowercase();
        if let Some(new_sw) = quick_map.get(&key) {
            // 在快速扫描中仍能找到 → 保留（使用最新信息）
            result.push((*new_sw).clone());
        } else {
            // 快速扫描中未找到 → 检查执行路径是否仍然存在
            let path = Path::new(&sw.exec_path);
            if path.exists() && sw.source != "registry" {
                // 路径仍存在 → 保留
                result.push(sw.clone());
            } else if sw.source == "common_paths" {
                // 来自深度扫描的软件，检查常见路径
                let fname = sw.exec_path.rsplit('\\').next().unwrap_or("");
                let fname_lower = fname.to_lowercase();
                if COMMON_EXES.contains(&fname_lower.as_str()) && path.exists() {
                    result.push(sw.clone());
                } else {
                    removed_count += 1;
                    log::info!("增量扫描: 移除已卸载软件 '{}'", sw.name);
                }
            } else {
                removed_count += 1;
                log::info!("增量扫描: 移除已卸载软件 '{}'", sw.name);
            }
        }
    }

    // 添加新发现的软件
    let result_names: std::collections::HashSet<String> = result.iter()
        .map(|sw| sw.name.to_lowercase())
        .collect();

    let mut added_count = 0;
    for sw in &quick_new {
        if !result_names.contains(&sw.name.to_lowercase()) {
            result.push(sw.clone());
            added_count += 1;
            log::info!("增量扫描: 发现新软件 '{}'", sw.name);
        }
    }

    if added_count > 0 || removed_count > 0 {
        log::info!("增量扫描: +{} / -{} 个软件", added_count, removed_count);
    }

    result.sort_by(|a, b| a.category.cmp(&b.category).then(a.name.cmp(&b.name)));
    save_software_cache(memory_path, &result);
    result
}

File: backend/core/software_scanner.py
import os
import re
import json
import subprocess
import winreg
from pathlib import Path
from datetime import datetime
from core.config import MEMORY_PATH

SOFTWARE_CACHE_FILE = os.path.join(MEMORY_PATH, "software_cache.json")
SOFTWARE_SCAN_FLAG = os.path.join(MEMORY_PATH, ".software_scanned")

CATEGORY_KEYWORDS = {
    "browser": ["browser", "chrome", "edge", "firefox", "safari", "opera", "chromium", "brave", "iexplore", "internet"],
    "editor": ["code", "sublime", "vim", "neovim", "notepad", "emacs", "atom", "brackets", "text", "ide"],
    "terminal": ["terminal", "cmd", "powershell", "windowsterminal", "conemu", "cmder", "hyper", "alacritty", "putty", "ssh", "warp"],
    "office": ["word", "excel", "powerpoint", "outlook", "onenote", "access", "libreoffice", "wps", "office"],
    "image": ["photoshop", "gimp", "krita", "paint", "illustrator", "figma", "sketch", "lightroom", "canvas", "photo", "draw", "inkscape", "corel"],
    "video": ["premiere", "after effects", "davinci", "final cut", "vegas", "kdenlive", "shotcut", "obs", "handbrake", "vlc", "mpv", "media player", "potplayer"],
    "audio": ["audacity", "ableton", "fl studio", "cubase", "logic", "reason", "studio one", "reaper", "garageband"],
    "development": ["git", "python", "node", "npm", "docker", "kubernetes", "vscode", "visual studio", "intellij", "pycharm", "webstorm", "goland", "clion", "eclipse", "android studio", "xcode", "jupyter", "anaconda", "cmake", "mingw", "cygwin", "gradle", "maven"],
    "database": ["mysql", "postgresql", "mongodb", "sqlite", "redis", "oracle", "sql server", "dbeaver", "sequel pro", "heidisql", "navicat", "pgadmin", "workbench"],
    "design": ["blender", "maya", "3ds max", "cinema 4d", "unity", "unreal", "godot", "fusion 360", "autocad", "solidworks", "sketchup", "rhino"],
    "communication": ["discord", "slack", "teams", "zoom", "skype", "telegram", "whatsapp", "wechat", "qq", "dingtalk", "lark", "feishu"],
    "utility": ["calculator", "calendar", "clock", "notes", "evernote", "notion", "obsidian", "roam", "1password", "lastpass", "dropbox", "onedrive", "google drive"],
    "game": ["steam", "epic", "battle.net", "origin", "ubisoft", "gog", "minecraft", "league", "dota", "counter-strike"],
    "music": ["spotify", "netease", "qq music", "apple music", "foobar", "winamp", "aimp", "musicbee"],
    "pdf": ["acrobat", "reader", "foxit", "sumatra", "pdf reader", "pdf viewer", "pdf-xchange"],
    "compression": ["winrar", "7-zip", "winzip", "peazip", "bandizip", "tar", "gzip"],
    "download": ["aria2", "motrix", "xdown", "idm", "internet download manager", "transmission", "qbittorrent", "utorrent", "thunder"],
}

EXCLUDED_NAMES = [
    "uninstall", "update", "setup", "install", "readme", "help",
    "documentation", "support", "feedback", "report", "get started",
    "what's new", "about", "license", "release notes",
]

EXCLUDED_DIRS = [
    "windows", "windows.old", "system32", "syswow64", "msocache",
    "$recycle.bin", "system volume information", "perflogs",
    "programdata\\microsoft", "appdata\\local\\temp", "appdata\\locallow",
    "appdata\\roaming\\microsoft", "common files",
]


class SoftwareInfo:
    def __init__(self, name: str, exec_path: str, icon_path: str = "",
                 category: str = "other", description: str = "",
                 source: str = "unknown"):
        self.name = name
        self.exec_path = exec_path
        self.icon_path = icon_path or exec_path
        self.category = category
        self.description = description or name
        self.source = source

    def to_dict(self):
        return {
            "name": self.name,
            "exec_path": self.exec_path,
            "icon_path": self.icon_path,
            "category": self.category,
            "description": self.description,
            "source": self.source,
        }

    @staticmethod
    def from_dict(d):
        return SoftwareInfo(
            name=d["name"],
            exec_path=d["exec_path"],
            icon_path=d.get("icon_path", d["exec_path"]),
            category=d.get("category", "other"),
            description=d.get("description", d["name"]),
            source=d.get("source", "cache"),
        )

    def __repr__(self):
        return f"[{self.category}] {self.name}"


def classify_software(name: str, exec_path: str = "") -> str:
    name_lower = name.lower()
    path_lower = exec_path.lower()

    combined = name_lower + " " + path_lower

    for cat, keywords in CATEGORY_KEYWORDS.items():
        for kw in keywords:
            if kw in combined:
                return cat
    return "other"


def is_excluded(name: str) -> bool:
    name_lower = name.lower().strip()
    for exc in EXCLUDED_NAMES:
        if exc in name_lower:
            return True
    return False


def scan_start_menu() -> list:
    found = []
    start_menu_dirs = [
        os.environ.get("ProgramData", "") + "\\Microsoft\\Windows\\Start Menu\\Programs",
        os.environ.get("APPDATA", "") + "\\Microsoft\\Windows\\Start Menu\\Programs",
        os.environ.get("ALLUSERSPROFILE", "") + "\\Microsoft\\Windows\\Start Menu\\Programs",
    ]

    seen = set()

    for base_dir in start_menu_dirs:
        if not os.path.isdir(base_dir):
            continue
        try:
            for root, dirs, files in os.walk(base_dir):
                skip = False
                for exc in EXCLUDED_DIRS:
                    if exc.lower() in root.lower():
                        skip = True
                        break
                if skip:
                    continue

                for f in files:
                    if not f.endswith(".lnk"):
                        continue
                    name = os.path.splitext(f)[0]
                    if is_excluded(name):
                        continue
                    lnk_path = os.path.join(root, f)

                    lower = name.lower()
                    if lower in seen:
                        continue
                    seen.add(lower)

                    category = classify_software(name, lnk_path)
                    found.append(SoftwareInfo(
                        name=name,
                        exec_path=lnk_path,
                        icon_path=lnk_path,
                        category=category,
                        description=f"从开始菜单发现: {name}",
                        source="start_menu",
                    ))
        except (PermissionError, OSError):
            continue

    return found


def scan_registry() -> list:
    found = []
    seen_names = set()

    registry_paths = [
        (winreg.HKEY_LOCAL_MACHINE, r"SOFTWARE\Microsoft\Windows\CurrentVersion\Uninstall"),
        (winreg.HKEY_LOCAL_MACHINE, r"SOFTWARE\WOW6432Node\Microsoft\Windows\CurrentVersion\Uninstall"),
        (winreg.HKEY_CURRENT_USER, r"SOFTWARE\Microsoft\Windows\CurrentVersion\Uninstall"),
    ]

    for hkey, key_path in registry_paths:
        try:
            key = winreg.OpenKey(hkey, key_path, 0, winreg.KEY_READ)
            try:
                i = 0
                while True:
                    try:
                        subkey_name = winreg.EnumKey(key, i)
                        i += 1
                        subkey = winreg.OpenKey(key, subkey_name)
                        try:
                            name, _ = winreg.QueryValueEx(subkey, "DisplayName")
                            if not name or is_excluded(name):
                                continue

                            lower = name.lower()
                            if lower in seen_names:
                                continue
                            seen_names.add(lower)

                            exec_path = ""
                            try:
                                exec_path, _ = winreg.QueryValueEx(subkey, "DisplayIcon")
                            except (FileNotFoundError, OSError):
                                pass
                            try:
                                if not exec_path:
                                    exec_path, _ = winreg.QueryValueEx(subkey, "InstallLocation")
                            except (FileNotFoundError, OSError):
                                pass

                            publisher = ""
                            try:
                                publisher, _ = winreg.QueryValueEx(subkey, "Publisher")
                            except (FileNotFoundError, OSError):
                                pass

                            category = classify_software(name, exec_path)
                            found.append(SoftwareInfo(
                                name=name,
                                exec_path=exec_path,
                                icon_path=exec_path,
                                category=category,
                                description=f"已安装: {name}" + (f" (by {publisher})" if publisher else ""),
                                source="registry",
                            ))
                        except (FileNotFoundError, OSError):
                            pass
                        finally:
                            winreg.CloseKey(subkey)
                    except (FileNotFoundError, OSError):
                        break
            finally:
                winreg.CloseKey(key)
        except (FileNotFoundError, OSError):
            continue

    return found


def scan_common_paths() -> list:
    found = []
    seen_names = set()

    common_dirs = [
        os.environ.get("ProgramFiles", "C:\\Program Files"),
        os.environ.get("ProgramFiles(x86)", "C:\\Program Files (x86)"),
        os.environ.get("LOCALAPPDATA", "") + "\\Programs",
        os.environ.get("LOCALAPPDATA", "") + "\\Microsoft\\WinGet\\Packages",
    ]

    common_exes = [
        "chrome.exe", "firefox.exe", "msedge.exe", "code.exe",
        "notepad++.exe", "sublime_text.exe", "terminal.exe",
        "wt.exe", "powershell.exe", "cmd.exe",
        "spotify.exe", "discord.exe", "slack.exe",
        "obs64.exe", "obs32.exe", "vlc.exe",
        "git-bash.exe", "git-cmd.exe",
        "python.exe", "pythonw.exe",
        "winrar.exe", "7zFM.exe",
        "sumatraPDF.exe",
    ]

    for base_dir in common_dirs:
        if not os.path.isdir(base_dir):
            continue
        try:
            for root, dirs, files in os.walk(base_dir):
                skip = False
                for exc in EXCLUDED_DIRS:
                    if exc.lower() in root.lower():
                        skip = True
                        break
                if skip:
                    continue
                for f in files:
                    if f.lower() not in common_exes:
                        continue
                    name = os.path.splitext(f)[0]
                    if is_excluded(name):
                        continue
                    lower = name.lower()
                    if lower in seen_names:
                        continue
                    seen_names.add(lower)

                    full_path = os.path.join(root, f)
                    category = classify_software(name, full_path)
                    found.append(SoftwareInfo(
                        name=name.capitalize(),
                        exec_path=full_path,
                        icon_path=full_path,
                        category=category,
                        description=f"在安装目录发现: {name}",
                        source="common_paths",
                    ))
        except (PermissionError, OSError):
            continue

    return found


def merge_software_lists(*lists):
    seen = {}
    merged = []

    for sw_list in lists:
        for sw in sw_list:
            key = sw.name.lower()
            if key not in seen:
                seen[key] = sw
                merged.append(sw)
            else:
                existing = seen[key]
                if not existing.exec_path and sw.exec_path:
                    existing.exec_path = sw.exec_path
                    existing.icon_path = sw.icon_path or existing.icon_path
                    existing.source = sw.source
                    if sw.category != "other":
                        existing.category = sw.category

    return merged


def scan_all_software() -> list:
    start_menu = scan_start_menu()
    registry = scan_registry()
    common = scan_common_paths()

    merged = merge_software_lists(start_menu, registry, common)

    categorized = {}
    for sw in merged:
        cat = sw.category
        if cat not in categorized:
            categorized[cat] = []
        categorized[cat].append(sw)

    result = []
    for cat in sorted(categorized.keys()):
        categorized[cat].sort(key=lambda x: x.name.lower())
        result.extend(categorized[cat])

    return result


def save_software_cache(software_list: list):
    os.makedirs(MEMORY_PATH, exist_ok=True)
    data = {
        "scanned_at": datetime.now().isoformat(),
        "count": len(software_list),
        "software": [sw.to_dict() for sw in software_list],
    }
    with open(SOFTWARE_CACHE_FILE, "w", encoding="utf-8") as f:
        json.dump(data, f, ensure_ascii=False, indent=2)

    with open(SOFTWARE_SCAN_FLAG, "w") as f:
        f.write(datetime.now().isoformat())

    print(f"[SoftwareScanner] 已缓存 {len(software_list)} 个软件到 {SOFTWARE_CACHE_FILE}")


def load_software_cache() -> list:
    if not os.path.exists(SOFTWARE_CACHE_FILE):
        return []
    try:
        with open(SOFTWARE_CACHE_FILE, "r", encoding="utf-8") as f:
            data = json.load(f)
        software = [SoftwareInfo.from_dict(d) for d in data.get("software", [])]
        print(f"[SoftwareScanner] 从缓存加载了 {len(software)} 个软件")
        return software
    except (json.JSONDecodeError, KeyError, Exception) as e:
        print(f"[SoftwareScanner] 缓存加载失败: {e}")
        return []


def is_software_scanned() -> bool:
    return os.path.exists(SOFTWARE_SCAN_FLAG)


def get_scan_time() -> str:
    if not os.path.exists(SOFTWARE_SCAN_FLAG):
        return ""
    try:
        with open(SOFTWARE_SCAN_FLAG, "r") as f:
            return f.read().strip()
    except Exception:
        return ""


def search_software(query: str, software_list: list, top_k: int = 10) -> list:
    query_lower = query.lower()

    scored = []
    for sw in software_list:
        score = 0
        if query_lower in sw.name.lower():
            score += 10
        if query_lower in sw.category:
            score += 5
        if query_lower in sw.description.lower():
            score += 3
        if sw.exec_path and query_lower in sw.exec_path.lower():
            score += 2

        for keyword in query_lower.split():
            if keyword in sw.name.lower():
                score += 2

        if score > 0:
            scored.append((score, sw))

    scored.sort(key=lambda x: -x[0])
    return [sw for _, sw in scored[:top_k]]


def launch_software(exec_path: str) -> dict:
    import subprocess
    try:
        if exec_path.lower().endswith(".lnk"):
            os.startfile(exec_path)
        else:
            subprocess.Popen([exec_path], shell=True)
        return {"success": True, "message": f"已启动: {os.path.basename(exec_path)}"}
    except Exception as e:
        return {"success": False, "message": f"启动失败: {str(e)}"}


def get_all_software() -> list:
    return load_software_cache()


def get_software_by_category(category: str) -> list:
    all_sw = load_software_cache()
    return [sw for sw in all_sw if sw.category == category]


def get_all_categories() -> dict:
    all_sw = load_software_cache()
    cats = {}
    for sw in all_sw:
        if sw.category not in cats:
            cats[sw.category] = []
        cats[sw.category].append(sw.name)
    return cats


def get_software_count() -> int:
    all_sw = load_software_cache()
    return len(all_sw)


def scan_and_cache() -> list:
    if is_software_scanned():
        print("[SoftwareScanner] 软件已扫描过，从缓存加载")
        cached = load_software_cache()
        if cached:
            return cached

    print("[SoftwareScanner] 开始扫描电脑上的软件...")
    software_list = scan_all_software()
    save_software_cache(software_list)

    total = len(software_list)
    categories = {}
    for sw in software_list:
        categories[sw.category] = categories.get(sw.category, 0) + 1

    print(f"[SoftwareScanner] 扫描完成! 共发现 {total} 个软件，涵盖 {len(categories)} 个分类")
    for cat, count in sorted(categories.items()):
        print(f"  {cat}: {count} 个")

    return software_list
